src/substrait/textplan/parser/SubstraitPlanTypeVisitor.cpp

```diff
@@ -0,0 +1,200 @@
+      } catch (...) {
+        errorListener_->addError(
+            ctx->getStart(), "Could not parse fixedchar length.");
+      }
+      type.mutable_fixed_char()->set_nullability(nullValue);
+      break;
+    }
+    case TypeKind::kVarchar: {
+      auto varChar =
+          reinterpret_cast<const ParameterizedFixedChar*>(&decodedType);
+      if (varChar == nullptr) {
+        break;
+      }
+      try {
+        int32_t length = std::stoi(varChar->length()->value());
+        type.mutable_varchar()->set_length(length);
+      } catch (...) {
+        errorListener_->addError(
+            ctx->getStart(), "Could not parse varchar length.");
+      }
+      type.mutable_varchar()->set_nullability(nullValue);
+      break;
+    }
+    case TypeKind::kFixedBinary:
+      type.mutable_fixed_binary()->set_nullability(nullValue);
+      break;
+    case TypeKind::kDecimal: {
+      auto dec = reinterpret_cast<const ParameterizedDecimal*>(&decodedType);
+      if (dec == nullptr) {
+        break;
+      }
+      try {
+        int32_t precision = std::stoi(dec->precision()->value());
+        int32_t scale = std::stoi(dec->scale()->value());
+        type.mutable_decimal()->set_precision(precision);
+        type.mutable_decimal()->set_scale(scale);
+      } catch (...) {
+        errorListener_->addError(
+            ctx->getStart(), "Could not parse decimal precision and scale.");
+      }
+      type.mutable_decimal()->set_nullability(nullValue);
+      break;
+    }
+    case TypeKind::kStruct: {
+      auto structure =
+          reinterpret_cast<const ParameterizedStruct*>(&decodedType);
+      for (const auto& t : structure->children()) {
+        *type.mutable_struct_()->add_types() = typeToProto(ctx, *t);
+      }
+      type.mutable_struct_()->set_nullability(nullValue);
+      break;
+    }
+    case TypeKind::kList: {
+      auto list = reinterpret_cast<const ParameterizedList*>(&decodedType);
+      *type.mutable_list()->mutable_type() =
+          typeToProto(ctx, *list->elementType());
+      type.mutable_list()->set_nullability(nullValue);
+      break;
+    }
+    case TypeKind::kMap: {
+      auto map = reinterpret_cast<const ParameterizedMap*>(&decodedType);
+      if (map->keyType() == nullptr || map->valueType() == nullptr) {
+        errorListener_->addError(
+            ctx->getStart(), "Maps require both a key and a value type.");
+        break;
+      }
+      *type.mutable_map()->mutable_key() = typeToProto(ctx, *map->keyType());
+      *type.mutable_map()->mutable_value() =
+          typeToProto(ctx, *map->valueType());
+      type.mutable_map()->set_nullability(nullValue);
+      break;
+    }
+    case TypeKind::kKindNotSet:
+      if (!insideStructLiteralWithExternalType(ctx)) {
+        errorListener_->addError(
+            ctx->getStart(), "Unable to recognize requested type.");
+      }
+      break;
+  }
+  return type;
+}
+
+bool SubstraitPlanTypeVisitor::insideStructLiteralWithExternalType(
```

Comment:
Can we add a comment explaining what is going on here?

Reply:
Done.